solana/indexer/balance_changes.go

```diff
@@ -0,0 +1,152 @@
+	tx *solana.Transaction,
+	blockTime time.Time,
+	trackedMints []string,
+	txLogger *zap.Logger,
+) error {
+	balanceChanges, err := extractBalanceChanges(meta, tx, trackedMints)
+	if err != nil {
+		return fmt.Errorf("failed to extract token balance changes: %w", err)
+	}
+	for acc, bal := range balanceChanges {
+		row := balanceChangeRow{
+			slot:           slot,
+			account:        acc,
+			balanceChange:  *bal,
+			signature:      tx.Signatures[0].String(),
+			blockTimestamp: blockTime,
+		}
+		err = insertBalanceChange(ctx, db, row)
+		if err != nil {
+			return fmt.Errorf("failed to insert balance change for account %s: %w", acc, err)
+		}
+
+		txLogger.Debug("balance change",
+			zap.String("account", row.account),
+			zap.String("mint", row.balanceChange.Mint),
+			zap.Uint64("balance", row.balanceChange.PostTokenBalance),
+			zap.Int64("change", row.balanceChange.Change),
+		)
+	}
+	return nil
+}
+
+type balanceChange struct {
+	Owner            string
+	Mint             string
+	PreTokenBalance  uint64
+	PostTokenBalance uint64
+	Change           int64
+}
+
+// Gets a map of account address to balance change from the given transaction.
+func extractBalanceChanges(meta *rpc.TransactionMeta, tx *solana.Transaction, trackedMints []string) (map[string]*balanceChange, error) {
+	balanceChanges := make(map[string]*balanceChange)
+
+	// Make a list of all accounts involved in the transaction
+	allAccounts, err := tx.Message.AccountMetaList()
+	if err != nil {
+		return nil, err
+	}
+
+	// Pre balances
+	for _, balance := range meta.PreTokenBalances {
+		isMintTracked := len(trackedMints) == 0 || slices.Contains(trackedMints, balance.Mint.String())
+		if !isMintTracked {
+			continue
+		}
+
+		acc := allAccounts[balance.AccountIndex].PublicKey
+		preBalance, err := strconv.ParseUint(balance.UiTokenAmount.Amount, 10, 64)
+		if err != nil {
+			return balanceChanges, err
+		}
+		owner := ""
+		if balance.Owner != nil {
+			owner = balance.Owner.String()
+		}
+
+		balanceChanges[acc.String()] = &balanceChange{
+			Owner:           owner,
+			Mint:            balance.Mint.String(),
+			PreTokenBalance: preBalance,
+		}
+	}
+
+	// Post balances and changes
+	for _, balance := range meta.PostTokenBalances {
+		isMintTracked := len(trackedMints) == 0 || slices.Contains(trackedMints, balance.Mint.String())
+		if !isMintTracked {
+			continue
+		}
+
+		acc := allAccounts[balance.AccountIndex].PublicKey
+		postBalance, err := strconv.ParseUint(balance.UiTokenAmount.Amount, 10, 64)
+		if err != nil {
+			return balanceChanges, err
+		}
+
+		b := balanceChanges[acc.String()]
+		if b == nil {
+			owner := ""
+			if balance.Owner != nil {
+				owner = balance.Owner.String()
+			}
+			b = &balanceChange{
+				Owner:           owner,
+				Mint:            balance.Mint.String(),
+				PreTokenBalance: 0,
+			}
+			balanceChanges[acc.String()] = b
+		}
+		b.PostTokenBalance = postBalance
+		b.Change = int64(b.PostTokenBalance) - int64(b.PreTokenBalance)
```

Comment:
Is this safe? Wouldn't you want to do the math first and then cast?

Reply:
yeah no this is definitely not safe.. good callout. Annoyingly these are both uint64 so there's really no way to fit the difference in an int64 without storing the sign separately. Maybe that's what we ought to do?

But yeah the logic could be something like "if a is greater than b, then a - b, else 0-int64(b - a)" to be a little safer?

wdyt

Reply:
Yeah I think that's reasonable. I've only ever used bignumber kinds of structs that do the math internally with special structs 😅